RPP Tensor Support - Fisheye on HOST and HIP

sampath1117 opened the pull request:
- Adds tensor support for Fog function optimized on AVX2 and HIP
- Adds test suite support

sampath1117 committed:
added initial api support for fisheye HIP kernel

sampath1117 committed:
added initial HIP support for PKD3-PKD3 variant

sampath1117 committed:
added ROI support

consolidated the code

sampath1117 committed:
added support for remaining layout variants

sampath1117 committed:
added few comments in the code

sampath1117 committed:
added initial api support for fisheye HOST kernel

sampath1117 committed:
added initial code for fishey src loc compute and support for U8 PKD3-PLN3 variant

sampath1117 committed:
added support for reamining layout variants in U8

sampath1117 committed:
added support for F16 bitdepth

sampath1117 committed:
added support for F32 and I8 bitdepths in HOST

sampath1117 committed:
Merge branch 'develop' into sr/opt_fisheye

sampath1117 committed:
added the versioning changes

sampath1117 committed:
added fisheye output image in doxygen

sampath1117 committed:
added golden outputs

HazarathKumarM committed:
Merge with develop

HazarathKumarM committed:
Merge remote-tracking branch 'develop' into sr/opt_fisheye

HazarathKumarM committed:
Fixes on the QA Failure on the HOST side

HazarathKumarM committed:
Merge branch 'develop' into opt_fisheye

HazarathKumarM committed:
Changes in Test suite

HazarathKumarM committed:
Minor changes due to merge conflits

Srihari-mcw commented on CHANGELOG.md:
```diff
@@ -2,12 +2,19 @@
 
 Full documentation for RPP is available at [https://rocm.docs.amd.com/projects/rpp/en/latest](https://rocm.docs.amd.com/projects/rpp/en/latest)
 
+## RPP 1.18.0 (unreleased)
```
Add the change under 1.9.10

Srihari-mcw commented on CMakeLists.txt:
```diff
@@ -43,7 +43,7 @@ endif()
 
 # RPP Version
-set(VERSION "1.9.10")
+set(VERSION "1.18.0")
```
Set version here to 1.9.10

HazarathKumarM committed:
Merge branch 'develop' into sr/opt_fisheye

Srihari-mcw commented on include/rpp_version.h:
```diff
@@ -39,8 +39,8 @@ extern "C" {
 #endif
 // NOTE: IMPORTANT: Match the version with CMakelists.txt version
 #define RPP_VERSION_MAJOR 1
-#define RPP_VERSION_MINOR 9
```
Here also retain version to 1.9.10

Srihari-mcw commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -120,6 +120,8 @@ const __m256 avx_p255 = _mm256_set1_ps(255.0f);
 const __m256 avx_p1op255 = _mm256_set1_ps(1.0f / 255.0f);
 const __m256 avx_p1op3 = _mm256_set1_ps(1.0f / 3.0f);
 const __m256 avx_p2op3 = _mm256_set1_ps(2.0f / 3.0f);
+const __m256 avx_pMinus1 = _mm256_set1_ps(-1);
```
Can we make this -1.0f?

Srihari-mcw commented on src/modules/cpu/kernel/fisheye.hpp:
```diff
@@ -0,0 +1,987 @@
+
+inline void compute_fisheye_src_loc_avx(__m256 &pDstY, __m256 &pDstX, __m256 &pSrcY, __m256 &pSrcX, __m256 &pHeight, __m256 &pWidth)
+{
+    __m256 pNormX, pNormY, pDist;
```
Add the AVX helper functions overall inside #if __AVX2__

Srihari-mcw commented on src/modules/cpu/kernel/fisheye.hpp:
```diff
@@ -0,0 +1,987 @@
+        }
+
+        // fisheye without fused output-layout toggle (NHWC -> NHWC)
+        if ((srcDescPtr->c == 3) && (srcDescPtr->layout == RpptLayout::NHWC) && (dstDescPtr->layout == RpptLayout::NHWC))
```
this must be else if

Srihari-mcw commented on src/modules/cpu/kernel/fisheye.hpp:
```diff
@@ -0,0 +1,987 @@
+                    {
+                        __m256i pRow;
+                        rpp_simd_load(rpp_generic_nn_load_i8pln1_avx, srcPtrTempChn, srcLocArray, invalidLoad, pRow);
+                        rpp_storeu_si64((__m128i *)(dstPtrTempChn), _mm256_castsi256_si128(pRow));
```
Can probably use reinterpret_cast<__m128i *>(dstPtrTempChn) here

Srihari-mcw commented on src/modules/hip/kernel/fisheye.hpp:
```diff
@@ -0,0 +1,299 @@
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    int4 srcRoi_i4 = *(int4 *)&roiTensorPtrSrc[id_z];
+    int width = (srcRoi_i4.z - srcRoi_i4.x) + 1;
+    int height =  (srcRoi_i4.w - srcRoi_i4.y) + 1;
```
Fix indentation

Srihari-mcw replied:
in all the functions

Srihari-mcw reviewed (changes requested):
Please address review comments. Thanks

HazarathKumarM committed:
Minor changes based on the review commands

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.14.1 to 1.15.0 in /docs/sphinx (#515)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.14.1 to 1.15.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.14.1...v1.15.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

HazarathKumarM committed:
Merge branch 'develop' into sr/opt_fisheye

HazarathKumarM committed:
Changes due to merge conflits

Srihari-mcw commented:
Moved to different PR - Closed
